In [2]:
import numpy as np
import pandas as pd

In [3]:
df = pd.read_pickle('/content/drive/MyDrive/orders_products_merged.pkl')

In [4]:
df.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  
0         Soda        77              7     9.0  Mid-range product  
1         Soda        77              7     9.0  Mid-range product  
2         Soda        77              7     9.0  Mid-range product  
3         Soda        77              7     9.0  Mid-range product  
4         Soda        77              7     9.0  Mid-range product

Step 2: Creating busiest_days for the two busiest and two slowest days of the week

In [14]:
result = []

for value in df["order_dow"]:
  if (value == 0) or (value ==1):
    result.append("Busiest days")
  elif (value == 3) or (value ==4):
    result.append("Slowest days")
  else:
    result.append("Regularly busy")

In [15]:
df['busiest_days'] = result

In [16]:
df.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

     busiest_days  
0  Regularly busy  
1    Slowest days  
2    Slowest days  
3    Slowest days  
4    Slowest days

Step 3: Counting the total number of busiest, regularly busy and slowest days in the busiest_days columns and comparing them with the numbers of the days in order_dow column to check for accuracy

In [18]:
df['order_dow'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_dow, dtype: int64

In [17]:
df['busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_days, dtype: int64

Busiest days = 6204182+5660230 = 11864412,
Slowest days = 3840534+3783802 = 7624336,
Regular days = 4496490+4213830+4205791 = 12916111

The values in the busiest_days columns are accurate

Step 4: Creating a new column called “busiest_period_of_day

In [19]:
df['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [20]:
result_2 = []
busiest_hours = [10, 11, 14, 15, 13, 12, 16, 9, 17, 8, 18, 19]
average_hours = [20, 7, 21, 22, 23]
slowest_hours = [6, 0, 1, 5, 2, 4, 3]


In [21]:
for value in df["order_hour_of_day"]:
  if value in busiest_hours:
    result_2.append("Most orders")
  elif value in slowest_hours:
    result_2.append("Fewest orders")
  else:
    result_2.append("Average orders")

In [24]:
df['busiest_period_of_day'] = result_2

In [25]:
df.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

     busiest_days busiest_period_of_day  
0  Regularly busy           Most orders  
1    Slowest days        Average orders  
2    Slowest days           Most orders  
3    Slowest days        Average orders  
4    Slowest days           Most orders

Step 5

In [26]:
df['busiest_period_of_day'].value_counts(dropna = False)

Most orders       27818650
Average orders     3699388
Fewest orders       886821
Name: busiest_period_of_day, dtype: int64

In [27]:
from google.colab import files
df.to_pickle('orders_products_merged_new.pkl') 
files.download('orders_products_merged_new.pkl')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>